spiketoolkit/postprocessing/postprocessing_tools.py

```diff
@@ -560,85 +719,147 @@ def compute_unit_pca_scores(recording, sorting, unit_ids=None, n_comp=3, by_elec
+                                                                           grouping_property=grouping_property,
+                                                                           dtype=dtype,
+                                                                           compute_property_from_recording=
+                                                                           compute_property_from_recording,
+                                                                           max_channels_per_waveforms=
+                                                                           max_channels_per_waveforms,
+                                                                           save_as_features=save_waveforms_as_features,
+                                                                           verbose=verbose, seed=seed, memmap=memmap,
+                                                                           return_idxs=True)
     else:
-        if verbose:
-            print("Computing waveforms")
-        waveforms, ind_list = get_unit_waveforms(recording, sorting, unit_ids, max_spikes_per_unit=max_spikes_per_unit,
-                                                 ms_before=ms_before, ms_after=ms_after,
-                                                 grouping_property=grouping_property, dtype=dtype,
-                                                 compute_property_from_recording=compute_property_from_recording,
-                                                 save_as_features=save_waveforms_as_features,
-                                                 verbose=verbose, seed=seed, return_idxs=True)
+        assert _spike_index_list is not None and _channel_list is not None, "Provide spike_index_list and " \
+                                                                            "channel_list with waveforms"
+        waveforms = _waveforms
+        spike_index_list = _spike_index_list
+        channel_list = _channel_list
+
+    # compute len of all waveforms (computed for all units)
+    n_waveforms = 0
+    for wf in waveforms:
+        n_spikes = len(wf)
+        n_waveforms += n_spikes
+    wf_shape = waveforms[0].shape
+
+    dtype = recording.get_dtype()
+    # prepare all waveforms
+    if memmap:
+        if by_electrode:
+            all_waveforms = np.memmap(tmp_folder / 'all_waveforms.raw', dtype=dtype,
+                                      shape=(n_waveforms * wf_shape[1], wf_shape[2]), mode='w+')
+        else:
+            all_waveforms = np.memmap(tmp_folder / 'all_waveforms.raw', dtype=dtype,
+                                      shape=(n_waveforms, wf_shape[1] * wf_shape[2]), mode='w+')
+    else:
+        if by_electrode:
+            all_waveforms = np.zeros((n_waveforms * wf_shape[1], wf_shape[2]), dtype=dtype)
+        else:
+            all_waveforms = np.zeros((n_waveforms, wf_shape[1] * wf_shape[2]), dtype=dtype)
```

Comment:
I don't understand the if by_electrode. because it do the same thing.

Reply:
no, by_electrode=True expands on the channels, by_electrode=False concatenates waveforms

Reply:
???

Reply:
the shape is different! 
by_electrode True: `(n_waveforms * wf_shape[1], wf_shape[2])`
by_electrode False: `(n_waveforms, wf_shape[1] * wf_shape[2])`

Reply:
oops. I miss the coma.
You should avoid a nested if here and make 2 diffrents if else. One for shape, one for memmap.

Reply:
removed. with the initialization feature of the allocate_array it's much cleaner now